File: hooks/lazydev-path-guard.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { platformPaths } from '../runtime/platform-policy.mjs';
import { nextAvailableArtifactName } from '../runtime/artifact-naming.mjs';

function norm(p) { return path.resolve(String(p || '')); }
function inside(target, root) {
  const a = norm(target), b = norm(root);
  return a === b || a.startsWith(b + path.sep);
}
function artifactDir() { return process.env.LAZYDEV_ARTIFACT_DIR ? norm(process.env.LAZYDEV_ARTIFACT_DIR) : norm(platformPaths().artifactDirectory); }
function loadContext() {
  const home = process.env.KIMI_CODE_HOME || path.join(process.env.HOME || process.cwd(), '.kimi-code');
  try { return JSON.parse(fs.readFileSync(path.join(home, 'lazydev-last-prompt.json'), 'utf8')); } catch { return {}; }
}
function explicitArtifactPrompt(prompt) {
  const p = String(prompt || '').toLowerCase();
  return /(save|export|download|generate|create|produce|write|file|artifact|deliverable)/u.test(p)
    && !/(edit|modify|refactor|fix|implement|source|repository|module|component|inside project|in the project)/u.test(p);
}
function getPath(input) {
  return String(input?.file_path || input?.path || input?.filename || '');
}
async function main() {
  let raw=''; process.stdin.setEncoding('utf8'); for await (const chunk of process.stdin) raw += chunk;
  let event={}; try { event=JSON.parse(raw); } catch { process.exit(0); }
  const input=event.tool_input || {};
  const target=getPath(input);
  if (!target) process.exit(0);
  const cwd=norm(event.cwd || process.cwd());
  const out=artifactDir();
  const resolved=norm(target);
  const ctx=loadContext();
  const implied=Boolean(ctx.task?.artifact) || explicitArtifactPrompt(ctx.prompt || '');
  const standaloneExt=/\.(?:html?|pdf|docx?|xlsx?|pptx?|zip|ahk|png|jpe?g|webp|gif|svg|csv|md|txt)$/iu.test(resolved);
  const workspaceRootFile=path.dirname(resolved) === cwd;
  const toolName=String(event.tool_name || '');
  const artifactSegment=path.basename(out);
  const pathSegments=resolved.split(path.sep).filter(Boolean);
  const looksLikeArtifactAlias=pathSegments.includes(artifactSegment) && !inside(resolved,out);
  // Misplaced standalone files are recoverable: the PostToolUse artifact
  // router moves successful Write/WriteFile outputs to the canonical path.
  // Keep this pre-tool hook focused on collision protection.
  if (implied && /^(?:Write|WriteFile)$/u.test(toolName) && inside(resolved, out) && fs.existsSync(resolved)) {
    const nextName = nextAvailableArtifactName(out, path.basename(resolved));
    process.stderr.write(`BLOCKED by LazyDev: the standalone deliverable already exists at ${resolved}. Keep the existing file untouched and retry Write with ${path.join(out, nextName)}. Do not claim success until that exact new path is verified.\n`);
    process.exit(2);
  }
  process.exit(0);
}
main().catch(() => process.exit(0));

File: hooks/lazydev-path-guard.py
#!/usr/bin/env python3
"""Prevent standalone artifacts from being written into a project root."""
from __future__ import annotations
import json, os, re, sys
from pathlib import Path

try:
    event = json.loads(sys.stdin.read() or "{}")
except Exception:
    raise SystemExit(0)
input_data = event.get("tool_input") or {}
target = str(input_data.get("file_path") or input_data.get("path") or input_data.get("filename") or "")
if not target:
    raise SystemExit(0)
cwd = Path(event.get("cwd") or os.getcwd()).resolve()
out = Path(os.getenv("LAZYDEV_ARTIFACT_DIR", Path.home() / ".local" / "share" / "lazydev" / "artifacts")).resolve()
resolved = Path(target).expanduser()
if not resolved.is_absolute():
    resolved = (cwd / resolved).resolve()
try:
    resolved.relative_to(out)
    raise SystemExit(0)
except ValueError:
    pass
ctx_path = Path(os.getenv("LAZYDEV_CONTEXT_DIR", Path.home() / ".lazydev")) / "last-prompt.json"
try:
    ctx = json.loads(ctx_path.read_text(encoding="utf-8"))
except Exception:
    ctx = {}
prompt = str(ctx.get("prompt", ""))
artifact = bool((ctx.get("task") or {}).get("artifact")) or bool(re.search(r"\b(save|export|download|generate|create|produce|artifact|deliverable)\b", prompt, re.I))
standalone = resolved.suffix.lower().lstrip(".") in {"html","htm","pdf","docx","xlsx","pptx","zip","png","jpg","jpeg","webp","gif","svg","csv","md","txt"}
if artifact and standalone and resolved.parent == cwd:
    print(f"BLOCKED by LazyDev: write standalone deliverables under {out}. Target was {resolved}.", file=sys.stderr)
    raise SystemExit(2)
raise SystemExit(0)
